### PR TITLE
Adding data base user attribute in EMF logs

### PR DESCRIPTION
**Pull request by ektabj**
*Issue #, if available:*

*Description of changes:*
- Added data base user attribute in EMF logs - `db.user`.
- Added unit test for data base user attribute
- Added contract-tests

Output of EMF log:
```
{
    "Environment": "ECS",
     ... ,
    "aws.remote.db.user": "myuser", <- Added Data base user
    "Error": {
        "Values": [
            0
        ],
        "Counts": [
            75574
        ]
       }
        ...
}
```

```
> Task :appsignals-tests:images:jdbc:jibDockerBuild                                                                                                                                            
                                                                                                                                                                                               
Containerizing application to Docker daemon as aws-appsignals-tests-jdbc-app...                                                                                                                
Base image 'public.ecr.aws/docker/library/amazoncorretto:17-alpine' does not use a specific image digest - build may not be reproducible                                                       
The base image requires auth. Trying again for public.ecr.aws/docker/library/amazoncorretto:17-alpine...                                                                                       
Using base image with digest: sha256:2122cb140fa94053abce343fb854d24f4c62ba3c1ac701882dce12980396b477                                                                                          
                                                                                                                                                                                               
Container entrypoint set to [java, -cp, @/app/jib-classpath-file, software.amazon.opentelemetry.Application]                                                                                   
                                                                                                                                                                                               
Built image to Docker daemon as aws-appsignals-tests-jdbc-app                                                                                                                                  
Executing tasks:                                                                                                                                                                               
[==============================] 100.0% complete 

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.                                                                                                    

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.                                                      

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.                                   

BUILD SUCCESSFUL in 16m 10s                                                                                                                                                                    
56 actionable tasks: 16 executed, 40 up-to-date    
```


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGenerator.java [hunk 1]
```diff
@@ -526,6 +529,12 @@ private static void setHttpStatus(SpanData span, AttributesBuilder builder) {
     }
   }
 
+  private static void setSpanKindForDBUser(SpanData span, AttributesBuilder builder) {
+    if (isKeyPresent(span, DB_USER)) {
+      builder.put(AWS_DB_USER, span.getAttributes().get(DB_USER));
+    }
+  }
+
```
Comment (thpierce) [inline awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGenerator.java:537]: Two notes:
1. `setSpanKind` doesn't make sense. Revise to `setRemoteDbUser`
2. Per agreed upon design, we should only set `db.user` if the span is a database span. Please add a check for `isDBSpan` before setting this value.

File: awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsAttributeKeys.java [hunk 2]
```diff
@@ -41,6 +41,8 @@ private AwsAttributeKeys() {}
   static final AttributeKey<String> AWS_REMOTE_RESOURCE_TYPE =
       AttributeKey.stringKey("aws.remote.resource.type");
 
+  static final AttributeKey<String> AWS_DB_USER = AttributeKey.stringKey("aws.remote.db.user");
```
Comment (thpierce) [inline awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsAttributeKeys.java:44]:
```suggestion
  static final AttributeKey<String> AWS_REMOTE_DB_USER = AttributeKey.stringKey("aws.remote.db.user");
```

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 3]
```diff
@@ -1029,6 +1030,38 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
     }
   }
 
+  @Test
+  void testDBUserAttribute() {
+    String dbUser = "test_user";
+    mockAttribute(DB_USER, dbUser);
+    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1037]: Please add a test for `SpanKind.SERVER` that asserts `AWS_DB_USER` attribute not present in `SERVICE_METRIC`

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 4]
```diff
@@ -1029,6 +1030,49 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
...
+
+  @Test
+  void testDBUserAttributeWithDifferentValues() {
+    String dbUser = "non_db_user";
+    mockAttribute(DB_USER, dbUser);
+    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
+
+    Attributes actualAttributes =
+        GENERATOR.generateMetricAttributeMapFromSpan(spanDataMock, resource).get(DEPENDENCY_METRIC);
+    assertThat(actualAttributes.get(AWS_REMOTE_DB_USER)).isEqualTo(dbUser);
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1062]: Does this or `testDBUserAttribute` actually work? The spanDataMock doesn't have `DB_SYSTEM`, `DB_OPERATION`, or `DB_STATEMENT` set, does it?

We need a test for `isDbSpan==True` and a test for `isDbSpan==False`

File: awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsSpanProcessingUtil.java [hunk 5]
```diff
@@ -226,6 +227,7 @@ private static String generateIngressOperation(SpanData span) {
   static boolean isDBSpan(SpanData span) {
     return isKeyPresent(span, DB_SYSTEM)
         || isKeyPresent(span, DB_OPERATION)
+        || isKeyPresent(span, DB_USER)
```
Comment (thpierce) [inline awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsSpanProcessingUtil.java:230]: I don't think we agreed to this. I think our definition of `isDbSpan` is still: `DB_SYSTEM`, `DB_OPERATION`, and/or `DB_STATEMENT`, why is this changing?

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/jdbc/JdbcContractTestBase.java
Comment (thpierce) [inline appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/jdbc/JdbcContractTestBase.java:1]: Why are we only changing JDBC tests? Is `db.user` not provided in any of the other contract tests?

Comment (ektabj) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/jdbc/JdbcContractTestBase.java:1]: I don't think `db.user` was meant for any other contract test except JDBC. Ref - [link](https://github.com/aws-observability/aws-otel-java-instrumentation/blob/main/appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/utils/SemanticConventionsConstants.java#L83)

Please correct me if I'm wrong.

File: awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsSpanProcessingUtil.java (hunk 5 above)
Comment (ektabj) [reply awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsSpanProcessingUtil.java:230]: Sure, will update this.

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java (hunk 4 above)
Comment (ektabj) [reply awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1062]: Will update..

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/jdbc/JdbcContractTestBase.java
Comment (thpierce) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/jdbc/JdbcContractTestBase.java:1]: Sorry I thought the mysql tests were already added, my mistake, this is fine.

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 6]
```diff
@@ -1029,6 +1030,84 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
...
+
+    assertThat(AwsSpanProcessingUtil.isDBSpan(spanDataMock)).isFalse();
+  }
+
+  @Test
+  public void testIsDbSpanFalse() {
+    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
+
+    assertThat(AwsSpanProcessingUtil.isDBSpan(spanDataMock)).isFalse();
+  }
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1109]: These tests are not really needed (we should already be testing `isDbSpan` logic in Util tests), but you are missing the one test you did need:
```suggestion
  @Test
  public void testDbUserPresentAndIsDbSpanFalse() {
    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
    mockAttribute(DB_USER, "DB user");
    Attributes actualAttributes =
        GENERATOR.generateMetricAttributeMapFromSpan(spanDataMock, resource).get(SERVICE_METRIC);
    assertThat(actualAttributes.get(AWS_REMOTE_DB_USER)).isNull();
  }
```

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 7]
```diff
@@ -1029,6 +1030,84 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
...
+
+    Attributes actualAttributes =
+        GENERATOR.generateMetricAttributeMapFromSpan(spanDataMock, resource).get(DEPENDENCY_METRIC);
+    assertThat(actualAttributes.get(AWS_REMOTE_OPERATION)).isEqualTo("db_operation");
+    assertThat(actualAttributes.get(AWS_REMOTE_DB_USER)).isEqualTo("db_user");
+  }
+
+  @Test
+  public void testDBUserAttributeAbsent() {
+    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1049]: You are testing `DbUserAttributeAbsentAndIsDbSpanFalse`, which is not as helpful:

```suggestion
    mockAttribute(DB_SYSTEM, "db_system");
    when(spanDataMock.getKind()).thenReturn(SpanKind.CLIENT);
```

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 8]
```diff
@@ -1029,6 +1030,84 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
...
+
+    Attributes actualAttributes =
+        GENERATOR.generateMetricAttributeMapFromSpan(spanDataMock, resource).get(DEPENDENCY_METRIC);
+    assertThat(actualAttributes.get(AWS_REMOTE_DB_USER)).isEqualTo("non_db_user");
+  }
+
+  @Test
+  public void testDBUserAttributeNotPresentInServiceMetricForServerSpan() {
+    String dbUser = "db_user";
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1071]: You are testing `DBUserAttributeNotPresentInServiceMetricForServerSpanAndIsDbSpanFalse`, which is not as helpful:
```suggestion
    mockAttribute(DB_SYSTEM, "db_system");
    String dbUser = "db_user";
```

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 9]
```diff
@@ -1029,6 +1030,84 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
...
+    Attributes actualAttributes =
+        GENERATOR.generateMetricAttributeMapFromSpan(spanDataMock, resource).get(DEPENDENCY_METRIC);
+    assertThat(actualAttributes.get(AWS_REMOTE_DB_USER)).isNull();
+  }
+
+  @Test
+  public void testDBUserAttributeWithDifferentValues() {
+    mockAttribute(DB_SYSTEM, "db_system");
+    mockAttribute(DB_OPERATION, "db_operation");
+    mockAttribute(DB_STATEMENT, "db_statement");
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1060]: Not blocking: Only need to set one of these.
```suggestion
    mockAttribute(DB_OPERATION, "db_operation");
```

File: awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java [hunk 10]
```diff
@@ -1029,6 +1030,84 @@ private void validateHttpStatusForNonLocalRootWithThrowableForClient(
     }
   }
 
+  @Test
+  public void testDBUserAttribute() {
+    mockAttribute(DB_SYSTEM, "db_system");
+    mockAttribute(DB_OPERATION, "db_operation");
+    mockAttribute(DB_STATEMENT, "db_statement");
```
Comment (thpierce) [inline awsagentprovider/src/test/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGeneratorTest.java:1037]: Not blocking: Only need to set one of these.
```suggestion
    mockAttribute(DB_STATEMENT, "db_statement");
```

**Review by thpierce — APPROVED** [review]

**Pull request merged**